Refactor prometheus integration tests

jzucker2 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change

I was initially inspired by a closed pull request here: https://github.com/home-assistant/core/pull/85553 to add some more labels but this pivoted into solely focusing on refactoring the prometheus integration tests to make them less brittle in anticipation of upcoming logic changes.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

jzucker2 committed:
Starting with a simple change

jzucker2 committed:
And trying again but actually adding the new area to this

jzucker2 commented:
Signed!

> Hi @jzucker2
> 
> It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant/core#113849).
> 
> Once you do that we will be able to review and accept this pull request.
> 
> Thanks!

Signed!

jzucker2 committed:
And that's getting interesting

knyar commented on homeassistant/components/prometheus/__init__.py:
```diff
@@ -345,6 +361,11 @@ def _labels(state: State) -> dict[str, Any]:
+            "area": state.attributes.get(ATTR_AREA_ID),
+            # TODO: how to get zone here?
+            # "zone": state.attributes.get(ATTR_AREA_ID),
+            "entity_id": state.object_id,
```
If this is called `object_id` internally, could you please use `object_id` for the label name as well?

knyar commented on homeassistant/components/prometheus/__init__.py:
```diff
@@ -345,6 +361,11 @@ def _labels(state: State) -> dict[str, Any]:
             "entity": state.entity_id,
             "domain": state.domain,
             "friendly_name": state.attributes.get(ATTR_FRIENDLY_NAME),
+            "area": state.attributes.get(ATTR_AREA_ID),
```
When the attribute is unset, the label value should be empty rather than the string `None`.
```suggestion
            "area": state.attributes.get(ATTR_AREA_ID) or "",
```

knyar commented on homeassistant/components/prometheus/__init__.py:
```diff
@@ -345,6 +361,11 @@ def _labels(state: State) -> dict[str, Any]:
+            # TODO: how to get zone here?
+            # "zone": state.attributes.get(ATTR_AREA_ID),
+            "entity_id": state.object_id,
+            "device_class": state.attributes.get(ATTR_DEVICE_CLASS),
```
Not sure if device class can be unset, but just in case it is:

```suggestion
            "device_class": state.attributes.get(ATTR_DEVICE_CLASS) or "",
```

knyar reviewed (commented):
Thank you.

Most of the [tests](https://github.com/home-assistant/core/blob/dev/tests/components/prometheus/test_init.py) check label values, so you will need to adjust those as well.

jzucker2 committed:
Wanted to add some small things to gitignore too

jzucker2 committed:
More metrics clean up

jzucker2 commented on homeassistant/components/prometheus/__init__.py:
```diff
@@ -345,6 +361,11 @@ def _labels(state: State) -> dict[str, Any]:
             "domain": state.domain,
             "friendly_name": state.attributes.get(ATTR_FRIENDLY_NAME),
+            "area": state.attributes.get(ATTR_AREA_ID),
+            # TODO: how to get zone here?
```
I just removed `zone` for now. We can address that in another pr (if the need arises, I'm not using it currently)

jzucker2 commented on .gitignore:
```diff
@@ -28,6 +28,9 @@ Icon
 
+# asdf
+.tool-versions
+
```
I feel kinda weird editing a root file, I could drop this if you'd rather

jzucker2 replied:
Removed from this pr

jzucker2 commented on .dockerignore:
```diff
@@ -4,6 +4,9 @@
 config
 docs
 
+# asdf
```
I feel kinda weird editing a root file, I could drop this if you'd rather

MartinHjelmare replied:
Please make a separate PR for this if it's needed.

jzucker2 replied:
Removed from this pr

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
The linter is harsh

jzucker2 commented on tests/components/prometheus/test_init.py:
```diff
@@ -143,9 +143,14 @@ async def test_setup_enumeration(
     body = await generate_latest_metrics(client)
     assert (
-        'homeassistant_sensor_temperature_celsius{domain="sensor",'
+        "homeassistant_sensor_temperature_celsius{"
```
The linter pushed this to be `"` which I found interesting

jzucker2 committed:
Need to adjust a ton of tests

jzucker2 committed:
I was finally able to commit

jzucker2 committed:
Trying to abstract metrics into a helper class

jzucker2 committed:
Fixed some tests at least

jzucker2 committed:
Making progress on tests

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Getting really close now

jzucker2 committed:
Only 1 or 2 tests left to fix

jzucker2 committed:
Only 1.5 tests left

jzucker2 committed:
That's more than enough for tonight

jzucker2 committed:
Got all the tests passing!

jzucker2 committed:
Another pass at test clean up

jzucker2 commented:
I fixed up all the tests! I'm not quite sure what the preferred process for test helpers is but I think it needs a little clean up. Tests *do* pass though

jzucker2 committed:
Fixed up all the tests, again

jzucker2 committed:
More clean up needed

jzucker2 committed:
Got device_class working just need to fix one test I broke

jzucker2 committed:
Got all the existing tests working!

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Refactored helpers into a separate file

jzucker2 committed:
I added some new tests! For the helpers, ironically

jzucker2 commented:
Hi! Thank you for the earlier feedback.

I added the labels (except for `zone`) and also updated the unit tests. It ended up being a huge effort to update the unit tests, they are unintentionally brittle for label changes. I refactored things into a helper class. I know that kind of reduces the legibility of a single test and abstracts a bit, but it should make other label changes much easier going forward.

As of result of finding the helper necessary, I did add some simple smoke tests for the helper.
I also do think the helper class could be cleaner, but it does work. And I was hoping to propose some other changes going forward specifically around prometheus and could clean it up then.

Let me know what you think! It's exciting to try and contribute back to a project I use so much.

jzucker2 committed:
Don't touch those files

jzucker2 committed:
Don't include that either

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Added my first real test

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

knyar commented on tests/components/prometheus/test_init.py:
```diff
@@ -137,15 +159,19 @@ async def test_setup_enumeration(
-        'homeassistant_sensor_temperature_celsius{domain="sensor",'
-        'entity="sensor.outside_temperature",'
-        'friendly_name="Outside Temperature"} 12.3' in body
+    MetricsTestHelper._perform_sensor_metric_assert(
+        "homeassistant_sensor_temperature_celsius",
+        "12.3",
+        "Outside Temperature",
+        "outside_temperature",
+        body,
+        device_class=SensorDeviceClass.TEMPERATURE,
```
Looking at the diff here, I think for an individual assertion the new version is less legible and clear.

The previous version was very simple: it confirmed that a particular string was present in a list.

In the new version, usage of positional arguments makes it difficult to understand what those arguments are. Perhaps the helpers should accept a `dict` with label keys & values, to make it more obvious what is being checked? E.g.:
```python
assert_metric_present(body, "homeassistant_sensor_temperature_celsius", 12.3, dict(
  entity="sensor.outside_temperature",
  domain="sensor",
  friendly_name="Outside Temperature",
  object_id="outside_temperature",
  device_class="temperature",
))
```

knyar commented on tests/components/prometheus/test_init.py:
```diff
@@ -162,16 +188,22 @@ async def test_view_empty_namespace(
+        "Radio Energy",
+        "radio_energy",
+        body,
+        device_class=SensorDeviceClass.POWER,
```
Could you please use strings for checking label values instead of referring to constants in other modules?

knyar commented on tests/components/prometheus/helpers.py:
```diff
@@ -0,0 +1,170 @@
+"""Helper methods for prometheus tests."""
+
+
+class MetricsTestHelper:
```
What's the purpose of this class? It seems like all of its methods can just be separate functions in this module.

knyar commented on tests/components/prometheus/helpers.py:
```diff
@@ -0,0 +1,170 @@
+            device_class=device_class,
+            metric_value=metric_value,
+        )
+        if positive_comparison:
```
It might be more convenient to provide separate assertion functions (e.g. `assert_metric_present` vs `assert_metric_absent`) rather than make the user set a bool.

knyar commented on tests/components/prometheus/helpers.py:
```diff
@@ -0,0 +1,170 @@
+        device_class=None,
+    ):
+        device_class_label_line = cls.get_device_class_label_line(device_class)
+        final_metric_value = f" {metric_value}" if metric_value else ""
```
Similar to my comment about `positive_comparison`, perhaps assertions without a value should be handled by separate functions.

knyar commented on tests/components/prometheus/test_init.py:
```diff
@@ -404,29 +446,30 @@ async def test_humidifier(
-        'humidifier_target_humidity_percent{domain="humidifier",'
-        'entity="humidifier.humidifier",'
-        'friendly_name="Humidifier"} 68.0' in body
+    MetricsTestHelper._perform_humidifier_metric_assert(
```
These device_class specific assertion methods seem like an unnecessary layer of abstraction. The `_perform_humidifier_metric_assert` is only used a handful of times here - it might be simpler to use a generic function, or have a helper function that's defined closer to where it's used.

knyar reviewed (changes requested):
> I added the labels (except for `zone`) and also updated the unit tests. It ended up being a huge effort to update the unit tests, they are unintentionally brittle for label changes. I refactored things into a helper class. I know that kind of reduces the legibility of a single test and abstracts a bit, but it should make other label changes much easier going forward.

Thank you!

I don't mind refactoring tests, but I would advise against combining such refactoring in a single PR with other changes.

At the moment it's hard to see the impact of the code change you have made, because all tests have changed as well. Coupling two changes like this makes reviewing (and potentially reverting) each of them more difficult.

May I suggest splitting test refactoring into a separate PR? I don't mind which of the two you'd like to discuss first. I shared some feedback on the specifics of the test refactoring here, and will be happy to discuss more in a separate issue or PR.

jzucker2 commented:
> > I added the labels (except for `zone`) and also updated the unit tests. It ended up being a huge effort to update the unit tests, they are unintentionally brittle for label changes. I refactored things into a helper class. I know that kind of reduces the legibility of a single test and abstracts a bit, but it should make other label changes much easier going forward.
> 
> Thank you!
> 
> I don't mind refactoring tests, but I would advise against combining such refactoring in a single PR with other changes.
> 
> At the moment it's hard to see the impact of the code change you have made, because all tests have changed as well. Coupling two changes like this makes reviewing (and potentially reverting) each of them more difficult.
> 
> May I suggest splitting test refactoring into a separate PR? I don't mind which of the two you'd like to discuss first. I shared some feedback on the specifics of the test refactoring here, and will be happy to discuss more in a separate issue or PR.

I'm willing to split this into 2 prs and keep test refactoring and logic changes separate. How about let's focus on test refactoring first, because simply adding "area" as a label to prometheus will break _every single test_ and I'd love to address that first (I'm also interested in a later pass to include the new tagging and labeling features in Home Assistant in prometheus metrics as well, but let's discuss that later).

Maybe I can make this PR the test refactoring PR since most of the work is around that. And I break out the logic changes into a follow up PR?

Thanks so much for the responses so far! I'm really looking forward to having these features in my Home Assistant set up.

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

knyar commented:
> Maybe I can make this PR the test refactoring PR since most of the work is around that. And I break out the logic changes into a follow up PR?

That sounds good! I've left a few comments about tests in my previous message. In general I think we should be careful about over-abstraction here: the purpose of these tests is to verify that exported metrics have the format that we expect, and to allow folks to understand and debug them I would suggest erring on the side of keeping them simple & explicit even if they seem verbose.

Do you have a list of goals for this refactoring or problems you've identified and want to fix? I think this might be helpful to describe (even if at a high level).

For example, if you find the current approach of hand-crafting full metric strings annoying, I think an alternative might be to parse metric output and examine it in a more structured way.

One thing that I find pretty inconvenient in our current testing approach is that most tests are broken into two functions: a fixture that creates Home Assistant entities, and the test function itself. These two functions are located in different parts of the same file, making it difficult to understand and modify a single test. Many fixtures are only used once. Perhaps moving fixtures closer to the test functions, or maybe combining them in some cases would be simpler? Not sure what's common here in Home Assistant ecosystem - I've not looked at how other components structure their testing, but I think it's worth examining and trying to be consistent with if we are making large changes here.

jzucker2 commented:
> > Maybe I can make this PR the test refactoring PR since most of the work is around that. And I break out the logic changes into a follow up PR?
> 
> That sounds good! I've left a few comments about tests in my previous message. In general I think we should be careful about over-abstraction here: the purpose of these tests is to verify that exported metrics have the format that we expect, and to allow folks to understand and debug them I would suggest erring on the side of keeping them simple & explicit even if they seem verbose.
> 
> Do you have a list of goals for this refactoring or problems you've identified and want to fix? I think this might be helpful to describe (even if at a high level).
> 
> For example, if you find the current approach of hand-crafting full metric strings annoying, I think an alternative might be to parse metric output and examine it in a more structured way.
> 
> One thing that I find pretty inconvenient in our current testing approach is that most tests are broken into two functions: a fixture that creates Home Assistant entities, and the test function itself. These two functions are located in different parts of the same file, making it difficult to understand and modify a single test. Many fixtures are only used once. Perhaps moving fixtures closer to the test functions, or maybe combining them in some cases would be simpler? Not sure what's common here in Home Assistant ecosystem - I've not looked at how other components structure their testing, but I think it's worth examining and trying to be consistent with if we are making large changes here.

So, long term goal: I wanted to add 2/3 labels here and ended up needing to adjust ~800 lines of tests. But the tests themselves are pretty good. So just hoping that we can have test helpers that are still legible and human-readable but flexible enough to make prometheus label changes reasonable going forward.

I do agree that I overcomplicated this first pass. I do see what you mean about fixtures and stuff.

I also saw your comment about following patterns elsewhere-- it seems this project is so big there's not exactly one way of doing things. But this approach did seem to make sense for the most part, except I definitely think I over-abstracted the refactor.

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

llamafilm commented on homeassistant/components/prometheus/__init__.py:
```diff
@@ -344,6 +359,9 @@ def _labels(state: State) -> dict[str, Any]:
             "entity": state.entity_id,
             "domain": state.domain,
             "friendly_name": state.attributes.get(ATTR_FRIENDLY_NAME),
```
Or shorter: `"friendly_name": state.attributes.get(ATTR_FRIENDLY_NAME, ""),`

llamafilm reviewed (changes requested):
I agree, it makes sense to change this PR to be only the refactor.  In that case, you should remove the lines about `ATTR_AREA_ID`, etc., and rename the PR.

frenck commented:
I've marked this PR a draft, as changes are requested that need to be processed.
Please un-draft it once it is ready for review again by clicking the **"Ready for review"** button.

Thanks! 👍 

../Frenck

[_Learn more about our pull request process._](https://developers.home-assistant.io/docs/review-process#prs-are-being-drafted-when-changes-are-needed)

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

emontnemery commented:
@jzucker2 do you plan to do split the PR as you mentioned here <https://github.com/home-assistant/core/pull/113849#issuecomment-2043612104> ?

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 commented:
> @jzucker2 do you plan to do split the PR as you mentioned here [#113849 (comment)](https://github.com/home-assistant/core/pull/113849#issuecomment-2043612104) ?

Yes! I was thinking maybe this should be the refactor the tests pr because most of the comments and changes really focus on that, especially compared to the logic. And I can split out logic changes into a follow up PR (that hopefully won't require this major amount of test refactoring again)

llamafilm commented:
As I said before, if this PR is meant to include only the refactor, then you should rename it and remove references to `ATTR_AREA_ID`, etc.

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

emontnemery commented:
> > @jzucker2 do you plan to do split the PR as you mentioned here [#113849 (comment)](https://github.com/home-assistant/core/pull/113849#issuecomment-2043612104) ?
> 
> Yes! I was thinking maybe this should be the refactor the tests pr because most of the comments and changes really focus on that, especially compared to the logic. And I can split out logic changes into a follow up PR (that hopefully won't require this major amount of test refactoring again)

That sounds like a plan!
I agree with this comment: https://github.com/home-assistant/core/pull/113849#issuecomment-2357024202

jzucker2 commented:
> > > @jzucker2 do you plan to do split the PR as you mentioned here [#113849 (comment)](https://github.com/home-assistant/core/pull/113849#issuecomment-2043612104) ?
> > 
> > 
> > Yes! I was thinking maybe this should be the refactor the tests pr because most of the comments and changes really focus on that, especially compared to the logic. And I can split out logic changes into a follow up PR (that hopefully won't require this major amount of test refactoring again)
> 
> That sounds like a plan! I agree with this comment: [#113849 (comment)](https://github.com/home-assistant/core/pull/113849#issuecomment-2357024202)

Awesome. I'll continue forward on that route. I should have a bunch of free time finally starting next week.

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Rolling back some logic changes to focus solely on tests

jzucker2 committed:
Curious what happens when I run the tests now

jzucker2 committed:
Getting closer to making things pass

jzucker2 committed:
Getting closer to a working pr now

jzucker2 committed:
Keeping up with test fixes

jzucker2 committed:
Getting much closer to something useful

jzucker2 committed:
Saving piecemeal

jzucker2 committed:
Getting closer to a final working version

jzucker2 committed:
Now that's an improvement

jzucker2 committed:
And moving a little forward

jzucker2 committed:
And now I'm really inching closer

jzucker2 committed:
Saving more complex test case fix

jzucker2 committed:
And now only 3 tests left

jzucker2 committed:
Getting close and only a few tests left

jzucker2 committed:
I think I'm close with only 1 test left

jzucker2 committed:
Does this mean the tests actually work now

jzucker2 committed:
Was not using the helper classes anymore

jzucker2 committed:
Now I'm really curious

jzucker2 committed:
Need to rename the recently renamed class

jzucker2 committed:
Was it really that easy? (No, it wasn't)

jzucker2 committed:
Is this finally enough

jzucker2 commented:
I removed all the logic changes and reworked the helpers to be more in line with the discussions and previous comments. I am ready for a review!

I added a few simple test for the helper class I created.

I chose to keep the helper class in this file because it was pretty small but happy to break it out or change things up if that would be preferred.

I am ready for a review!

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Also added another full percentage point of tests

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

knyar commented on tests/components/prometheus/test_init.py:
```diff
@@ -87,6 +87,71 @@
+    domain: str
+    friendly_name: str
+    object_id: str
+    metric_value: Any | None = None
```
Note that metric values in Prometheus are always floats.

jzucker2 replied:
Rewritten. That's a good point.

knyar commented on tests/components/prometheus/test_init.py:
```diff
@@ -87,6 +87,71 @@
+    friendly_name: str
+    object_id: str
+    metric_value: Any | None = None
+    mode: str | None = None
+    state: str | None = None
+    action: str | None = None
```
As you noticed, not all metrics currently have these labels, and we could have metrics in the future that use some other labels.

I would advise against hardcoding them like this. Might be better to store them in a dict perhaps?

jzucker2 replied:
I had liked the idea of typing, but if you'd prefer a dict then I'll go with that.

knyar commented on tests/components/prometheus/test_init.py:
```diff
@@ -87,6 +87,71 @@
+        )
+
+
+def _assert_metric_present(body, metric_info: EntityTestMetricInfo) -> None:
+    """Assert metrics text body contains metric for entity."""
+    assert metric_info.get_full_metric_string() in body
+
+
+def _assert_metric_not_present(body, metric_info: EntityTestMetricInfo) -> None:
+    """Assert metrics text body does not contain metric for entity."""
+    assert metric_info.get_full_metric_string() not in body
```
I wonder moving these to be methods on the Metric class would make tests more readable?

jzucker2 replied:
Great idea! Done!

knyar commented on tests/components/prometheus/test_init.py:
```diff
@@ -87,6 +87,71 @@
+        )
+
+    @property
+    def entity(self) -> str:
+        """Generate entity_id from components."""
+        return f"{self.domain}.{self.object_id}"
```
I am not sure that encoding label transformations in the testing helper is a good idea. The goal of the test is to confirm that the metric content matches expectations, and it's better to have the test be explicit (even if verbose) rather than hard-code additional logic in the test.

If we expect the metric to have an `entity` label, we should probably pass an `entity` value to these helpers.

jzucker2 replied:
I rewrote as you suggested. Considering that we already pass in `object_id` and `domain` to pretty much every test, also including `entity` means a lot of repetition. But is rewritten as you suggested.

knyar reviewed (commented):
Thanks for tidying this up. I have left a few comments, and also wanted to provide an example of what I think a slightly different version of this helper could look like as a result:

Since in some cases we care about values, and in some cases we don't, perhaps we could split the helper into a "metric" and "metric with value":

```python
from prometheus_client.utils import floatToGoString

class EntityMetric:
    """Represents a Prometheus metric for a Home Assistant entity."""

    metric_name: str
    labels: dict[str, str]

    def __init__(self, metric_name: str, **kwargs: dict[str, str]) -> None:
        """Create a new EntityMetric based on metric name and labels."""
        self.metric_name = metric_name
        self.labels = kwargs

        # Labels that are required for all entities.
        for labelname in ("domain", "friendly_name", "entity"):
            assert labelname in self.labels
            assert self.labels[labelname] != ""

    def withValue(self, value: float):
        """Return a metric with value."""
        return EntityMetricWithValue(self, value)

    @property
    def _metric_name_string(self) -> str:
        """Return a full metric name as a string."""
        labels = ",".join(
            f'{key}="{value}"' for key, value in sorted(self.labels.items())
        )
        return f"{self.metric_name}{{{labels}}}"

    def assertInBody(self, body: list[str]):
        """Assert that this metric exists in the provided Prometheus output."""
        assert any(line.startswith(self._metric_name_string) for line in body)

    def assertNotIn(self, body: list[str]):
        """Assert that this metric does not exist in Prometheus output."""
        return not self.assertInBody(body)


class EntityMetricWithValue(EntityMetric):
    """Represents a Prometheus metric with a value."""

    value: float

    def __init__(self, metric: EntityMetric, value: float) -> None:
        """Create a new metric with a value based on a metric."""
        super().__init__(metric.metric_name, **metric.labels)
        self.value = value

    @property
    def _metric_string(self) -> str:
        """Return a full metric string."""
        value = floatToGoString(self.value)
        return f"{self._metric_name_string} {value}"

    def assertInBody(self, body: list[str]):
        """Assert that this metric exists in the provided Prometheus output."""
        assert self._metric_string in body
```

Within a test, using the helper might look like this (taking `test_disabling_entity` as an example):

```python
    EntityMetric(
        metric_name="state_change_total",
        domain="sensor",
        entity="sensor.outside_temperature",
        friendly_name="Outside Temperature",
    ).withValue(1).assertInBody(body)

    EntityMetric(
        metric_name="state_change_created",
        domain="sensor",
        entity="sensor.outside_temperature",
        friendly_name="Outside Temperature",
    ).assertInBody(body)
```

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Trying to clean things up a bit more

jzucker2 committed:
Now how does this look?

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Just a little more clean up

jzucker2 committed:
Added a few more tests for the new helper functions

jzucker2 committed:
Last pass on much better tests for this

jzucker2 committed:
Oops, forgot to remove redundant tests

jzucker2 committed:
Fix the fixtures

knyar commented on tests/components/prometheus/test_init.py:
```diff
@@ -87,6 +88,64 @@
+        )
+        return f"{self.metric_name}{{{labels}}}"
+
+    def assertInBody(self, body: list[str]):
+        """Assert that this metric exists in the provided Prometheus output."""
+        assert any(line.startswith(self._metric_name_string) for line in body)
+
+    def assertNotInBody(self, body: list[str]):
+        """Assert that this metric does not exist in Prometheus output."""
+        assert self._metric_name_string not in body
```
Note that as written assertNotInBody will never match, because Prometheus output always includes values.

Perhaps add an internal method for the actual check (`_in_metrics`), and then use that from both assert functions? This will ensure that the logic always matches, and will allow `EntityMetricWithValue` to only override `_in_metrics`, using assert methods from the parent class.

Also a few cosmetic suggestions (sorry for not thinking about these earlier):
* it's not very intuitive what `body` refers to, since this is so far away from the code that passes HTTP response body to these methods. Let's call it `metrics` instead? Or `metric_response` if you want to be more explicit?
* camelCase is not very idiomatic in Python, I suspect it's better use underscores (surprised that there is no linter checking this?)

```suggestion
    def _in_metrics(self, metrics: list[str]) -> bool:
        """Report whether this metric exists in the provided Prometheus output."""
        return any(line.startswith(self._metric_name_string) for line in body)

    def assert_in_metrics(self, metrics: list[str]):
        """Assert that this metric exists in the provided Prometheus output."""
        assert self._in_metrics(metrics)

    def assert_not_in_metrics(self, metrics: list[str]):
        """Assert that this metric does not exist in Prometheus output."""
        assert not self._in_metrics(metrics)
```

knyar commented on tests/components/prometheus/test_init.py:
```diff
@@ -178,17 +403,19 @@ async def test_view_empty_namespace(
+        domain="sensor",
+        friendly_name="Radio Energy",
+        entity="sensor.radio_energy",
+    ).withValue(1.0).assertInBody(body)
```
You don't need to change this, but I don't think you needed to append `.0` to all integer values.

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Getting closer to something decent, I hope

jzucker2 committed:
Another pass on the formatting of the number 1

jzucker2 committed:
And yet another pass on these tests

jzucker2 commented on tests/components/prometheus/test_init.py:
```diff
@@ -250,23 +486,26 @@ async def test_sensor_without_unit(
+        entity="sensor.text",
+    ).withValue(0.0).assert_not_in_metrics(body)
 
-    assert (
```
I think this test was actually broken because the metric is present in the `body` but it is written as `0.0` not `0` so that might be why it fails in upstream? I swapped it in my PR and am open to suggestions around this. I'm not quite sure what we are testing here, tbh

knyar replied:
I am addressing this in #127262 - would you mind reviewing it please? I don't mind if this PR gets merged first, will be happy to rebase mine.

jzucker2 commented:
Updated everything again and the tests are still passing, though I had to amend a single test and I left a note as to the reason

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Tests cleaned up a bit more

knyar commented on tests/components/prometheus/test_init.py:
```diff
@@ -250,23 +515,26 @@ async def test_sensor_without_unit(
+        domain="sensor",
+        friendly_name="Text",
+        entity="sensor.text",
+    ).withValue(0.0).assert_not_in_metrics(body)
```
The `assert_not_in_metrics` method only exists on `EntityMetric`, which I think is good - if we don't want a metric to exist, we don't care what value it has.

But it means that the value passed here is unused, and not needed.

```suggestion
    ).assert_not_in_metrics(body)
```

jzucker2 replied:
I know this might be silly. Should I add a simple test for each of the `assert_in_metrics()` and `assert_not_in_metrics()` helpers? I added some simple tests for everything else.

knyar replied:
Adding a test would definitely be good, and help prevent possible regressions when these helper classes are changed in the future.

jzucker2 replied:
I added tests for the helpers for both classes, both for `assert_in_metrics` and `assert_not_in_metrics`

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Minor updates as suggested

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Another pass on assert with metrics helper

jzucker2 committed:
Now this is fully tested

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

jzucker2 committed:
Merge branch 'dev' into add-more-prometheus-labels

emontnemery reviewed (approved):
LGTM, nice first PR @jzucker2 🎉

MartinHjelmare commented on tests/components/prometheus/test_init.py:
```diff
@@ -87,6 +88,77 @@
 PROMETHEUS_PATH = "homeassistant.components.prometheus"
 
 
+class EntityMetric:
```
These helpers make the tests harder to read. Instead of just being able to read the tests I need to go look at this helper and try to understand how it works. It's not helpful in a project like Home Assistant with thousands of integrations if every integration designs its own test helpers and moves assertions to those.
